apps/sms/views/conversation/js/conversation.js

```diff
@@ -267,48 +267,44 @@ var ConversationView = {
     this.onMessageTypeChange = this.onMessageTypeChange.bind(this);
   },
 
-  onVisibilityChange: function conv_onVisibilityChange(e) {
-    // If we leave the app and are in a thread or compose window
-    // save a message draft if necessary
-    if (document.hidden) {
-      // Auto-save draft if the user has entered anything
-      // in the composer or into To field (for composer panel only).
-      var isAutoSaveRequired = false;
+  onVisibilityChange: function conv_onVisibilityChange() {
+    // If we leave the app and are in a thread or compose window update a
+    // message draft if necessary.
+    var isDraftUpdateNeeded = document.hidden &&
+      (Navigation.isCurrentPanel('composer') ||
+      Navigation.isCurrentPanel('thread'));
 
-      if (Navigation.isCurrentPanel('composer')) {
-        isAutoSaveRequired = !Compose.isEmpty() ||
-          !!ConversationView.recipients.length;
-      } else if (Navigation.isCurrentPanel('thread')) {
-        isAutoSaveRequired = !Compose.isEmpty();
-      }
-
-      if (isAutoSaveRequired) {
-        ConversationView.saveDraft(true /* preserveDraft */);
-      }
+    if (isDraftUpdateNeeded) {
+      this.updateDraft();
     }
   },
 
   /**
-   * Executes when the header 'action' button is tapped.
-   *
-   * @private
-   */
-  onHeaderAction: function conv_onHeaderAction() {
-    this.backOrClose();
-  },
-
-  /**
    * We always go back to the previous pane, unless we're in an activity, then
    * in selected cases we exit the activity.
    *
    * @private
    */
   backOrClose: function conv_backOrClose() {
-    var inActivity = ActivityClient.hasPendingRequest();
-    var isComposer = Navigation.isCurrentPanel('composer');
-    var isThread = Navigation.isCurrentPanel('thread');
-    var action = inActivity && (isComposer || isThread) ? 'close' : 'back';
-    this[action]();
+    this.stopRendering();
+
+    var backOrClosePromise;
+    if (ActivityClient.hasPendingRequest()) {
+      backOrClosePromise = this.updateDraft().then(
+        () => ActivityClient.postResult()
+      );
+    } else {
+      // We're waiting for the keyboard to disappear before animating back.
+      backOrClosePromise = this._ensureKeyboardIsHidden().then(() => {
```

Comment:
Hmm, interesting, I actually didn't want to wait just because it was so before, but I don't really  see any strong reason to not return now  :) Am I missing reason why we should not return ```toPanel``` here?